FR-5402 - ubuntu-image classic always passes `--vcs=auto` to germinate even when told not to

upils opened the pull request:
Fixing https://bugs.launchpad.net/ubuntu-image/+bug/2021411

sil2100 reviewed (changes requested):
Sadly, I know you already did work for this fix here, but I will have to indeed object to the renaming of `vcs` to `no-vcs`. One thing is what you mentioned in the LP bug already: backwards compatibility. I wouldn't want us to fix something by changing the schema per-se. Second, the schema so far doesn't really have any 'negative keys', aka. this would be the first inverse field in the yaml file. And I don't think it's very intuitive in this case.

It is indeed a troublesome feature of the boolean in this case, and I think it's a shame that the schema handling does not really handle fields being unassigned in some special way (like via a separate flag somewhere). Potentially this problem can appear for any boolean value, and I don't think we should build out schema against limitations of the language/framework we're using.

I know you mentioned a pointer approach, which I guess could be doable - we could design it in such a way to use it in any other future booleans if the need appears. I was personally even thinking of using a string here, but that seems a bit worse? Anyway, could you take a look at it so that we don't have to change the schema? Thank you!

upils commented:
Fixed it using pointers to booleans. I also updated the definition of `KeepEnabled` even though it is not used. I understand this is to keep a kind of compatibility with *crafts configuration, but should we removing it?

sil2100 commented on internal/helper/helper_test.go:
```diff
@@ -117,3 +117,177 @@ func TestFailedRestoreResolvConf(t *testing.T) {
+	A []*S4
+}
+
+func TestSetDefaults(t *testing.T) {
```
Nice!

sil2100 commented on internal/imagedefinition/image_definition.go:
```diff
@@ -97,7 +97,7 @@ type PPA struct {
 	Auth        string `yaml:"auth"         json:"Auth,omitempty"        jsonschema:"pattern=^[a-zA-Z0-9_.+-]+:[a-zA-Z0-9]+$"`
 	Fingerprint string `yaml:"fingerprint"  json:"Fingerprint,omitempty"`
-	KeepEnabled bool   `yaml:"keep-enabled" json:"KeepEnabled"           default:"true"`
+	KeepEnabled *bool  `yaml:"keep-enabled" json:"KeepEnabled"           default:"true"`
```
As for this, I'll mention more about it in my main review comment, but I missed that it's not used! I was sure we used it already!

sil2100 reviewed (approved):
This looks good. And I like the additional improvements to the testing story. This feels good for merging.

As for `KeepEnabled` - this is a bug/missing-feature if it's not used. I thought we did handle it properly. Might be that it got missed due to deadlines. Basically it means whether the enabled PPA should be kept enabled on the built system after package installation. Let me create a card for it.

sil2100 commented:
Ok, before merging, could you maybe remove the boolean handling though?

upils commented on internal/helper/helper.go:
```diff
@@ -123,36 +127,54 @@ func SetDefaults(needsDefaults interface{}) error {
+				case reflect.Slice:
+					defaultValues := strings.Split(defaultValue, ",")
+					field.Set(reflect.ValueOf(defaultValues))
+				case reflect.Bool:
```
I agree that is unreliable. I will error out when trying to set a default on a bool. I will also add a specific test case on the image definition to make sure we never add a bool with a default. Without this test we risk letting it pass and users will face an error they cannot do anything about.

sil2100 reviewed (approved):
Excellent!

sil2100 commented:
hm, linting failed, but that doesn't seem to be a real issue?

upils commented:
Yeah, just a minor linting error about import order. This is fixed.
